client/cspaillier.go

```diff
@@ -50,11 +53,7 @@ func (c *CSPaillierClient) Run() error {
 		return err
 	}
 
-	if err := c.close(); err != nil {
-		return err
-	}
-
-	return nil
+	return c.closeStream()
```

Comment:
As discussed offline, I suggest we use `defer` for all the clients. This way the code will become even more readable. The downside is that the possible error that comes from stream closing won't bubble up, but I'm not sure if this is a big problem.